### PR TITLE
[IMP] account: account owl migration

### PR DESCRIPTION
**Pull request by h4818**
Migrate `account` widgets and views to owl

## Field Widgets:
payment
account_type_selection
open_move_widget
section_and_note_one2many
section_and_note_text

## Views:
account_tree
account_documents_kanban
account_dashboard_kanban
account_move_form

Web:
fileuploader callback after multiple files are uploaded

### CODE REVIEW

File: addons/account/views/account_move_views.xml
Comment (h4818) [inline addons/account/views/account_move_views.xml:808]: I think this has been resolved, and no need to remove the placeholder

File: addons/account/static/src/js/bills_upload.js
Comment (Levizar) [inline addons/account/static/src/js/bills_upload.js:78]: why `!!uploadInput` instead of `uploadInput` ?

Comment (h4818) [reply addons/account/static/src/js/bills_upload.js:78]: [] is truthy - so bang bang

Comment (Levizar) [reply addons/account/static/src/js/bills_upload.js:78]: Yes but what does `!!` change to that? If you do `Boolean([])` or `!![]`, you get the same result.

Comment (h4818) [reply addons/account/static/src/js/bills_upload.js:78]: Boolean(v) and !!v will produce the same result - just a matter of preference.

Comment (sdegueldre) [reply addons/account/static/src/js/bills_upload.js:78]: `uploadInput` is not an array, maybe you meant to do `!!files`? But that would always be truthy anyways because as noted `[]` is truthy. Maybe you want `files.length > 0`?

`uploadInput` will be an `HTMLElement` or null. We also would not typically coerce values to booleans unless we have a good reason to do so (eg we want to call a function with the value in question and want to make sure we're not leaking object references)

File: addons/account/static/src/js/bills_upload.js
Comment (sdegueldre) [inline addons/account/static/src/js/bills_upload.js:50]: Why are we using a class to hide the dropzone instead of just not rendering it with a t-if when it's not needed?
It also seems like this component has basically no behaviour and is just a template, so you could just t-call the template instead of using a component.

File: addons/account/static/src/js/bills_upload.js
Comment (sdegueldre) [inline addons/account/static/src/js/bills_upload.js:57]: This is a mixin. Mixins have issues with regards to composability and they clobber the namespace. A hook can solve the same problems without these drawbacks. Come find me if it's not obvious to you how to convert this to a hook. It also looks like a good bit of this logic could be moved to the dropzone component itself, which would also allow you to use a ref instead of querying the DOM lower in this file.

File: addons/account/static/src/js/mail_activity.js
Comment (sdegueldre) [inline addons/account/static/src/js/mail_activity.js:24]: We no longer use underscore prefixes for private methods because all methods on owl components are private by default and nobody except owl itself should be handling component references that would allow them to call these methods.

As for method names, if you prefix the method name with "on", the rest of the name should describe the event that happened (eg `onOpenButtonClick`), though in most cases you should simply give the method a descriptive name that doesn't start with `on` which will make reading both the js and the xml of the component easier (eg `openActivity`

File: addons/account/static/src/js/mail_activity.js
Comment (sdegueldre) [inline addons/account/static/src/js/mail_activity.js:36]: you can use the .prevent event handler suffix top preventDefault directly in xml

File: addons/account/static/src/js/mail_activity.js
Comment (sdegueldre) [inline addons/account/static/src/js/mail_activity.js:50]: Is this supposed to stay?

File: addons/account/static/src/js/mail_activity.js
Comment (sdegueldre) [inline addons/account/static/src/js/mail_activity.js:51]: Naming

File: addons/account/static/src/js/open_move_widget.js
Comment (sdegueldre) [inline addons/account/static/src/js/open_move_widget.js:7]: Why does this extend CharField? It uses none of the char field behaviour.

File: addons/account/static/src/js/section_and_note_fields_backend.js
Comment (sdegueldre) [inline addons/account/static/src/js/section_and_note_fields_backend.js:25]: Prefer template literals to string concatenation, it helps legibility.

**Review by sdegueldre — COMMENTED** [review]
Couldn't finish the entire review today but I'll already post these comments so you can get started on it. Work is solid but lots of code = lots of small things to change. Thank you for your efforts!

File: addons/account/static/src/js/account_payment_field.js
Comment (h4818) [reply addons/account/static/src/js/account_payment_field.js:46]: This payment popper is heavily inspired by the `many2many_tags_field` color picker, which uses the usePopover hook. Visibility doesn't seem to be available through a prop or state.

File: addons/account/static/src/js/account_payment_field.js
Comment (h4818) [reply addons/account/static/src/js/account_payment_field.js:57]: I guess I should have questions the other implementation of usePopover - you are right. After trying it, I can just reference the PopoverComponent directly without specifying the `components` on the constructor

File: addons/account/static/src/js/account_payment_field.js
Comment (h4818) [reply addons/account/static/src/js/account_payment_field.js:92]: https://github.com/odoo/odoo/pull/96672/commits/549e5705e290ecfd1cd1f4eefaed65788c0deaa5#diff-f1438a3721fcbfbf802f24b32b0e2baa5439ea42f56f63ff4b81cd3fc5d52ec1R62

it's a bound function that is given to the popover as a prop

Comment (sdegueldre) [reply addons/account/static/src/js/account_payment_field.js:92]: In that case I don't think it can be an action button, seem good as is :+1:

File: addons/account/static/src/js/account_payment_field.js
Comment (sdegueldre) [reply addons/account/static/src/js/account_payment_field.js:46]: You can pass an onClose function to popover.add, which lets you track that (eg, pass `onClose = () => {this.popoverCloseFn = null}` and use the truth value of the closeFn as your indicator that the popover is already closed). This also gets rid of the need to have two variables with similar and confusing name (this.closePopover and this.popoverCloseFn) as the callback is invoked regardless of the close reason (user click outside or programmatic close)

File: addons/account/static/src/js/bills_upload.js
Comment (h4818) [reply addons/account/static/src/js/bills_upload.js:50]: if I use a `t-if` I get a "Component is destroyed" when using it on the kanbanRecord https://github.com/odoo/odoo/pull/96672/commits/5ee56d246f39d027053adb1434601668c8591d15#diff-f48f436b6c64a5ab4918a6f53fdb2584c42427fe29f56b7689d607c43e51481cR41

File: addons/account/static/src/js/section_and_note_fields_backend.js
Comment (Levizar) [inline addons/account/static/src/js/section_and_note_fields_backend.js:23]: Do we support the `get` JS feature?
I think remember Sam told me we couldn't use that because of browser compatibility

File: addons/account/static/src/js/tax_totals.js
Comment (Levizar) [inline addons/account/static/src/js/tax_totals.js:119]: get

File: addons/account/static/src/js/tax_totals.js
Comment (Levizar) [inline addons/account/static/src/js/tax_totals.js:124]: get

File: addons/account/static/src/js/vat_activity.js
Comment (Levizar) [inline addons/account/static/src/js/vat_activity.js:12]: Shouldn't we put a hook to format the data onPropsWillUpdate as you did somewhere else?

File: addons/account/static/src/js/section_and_note_fields_backend.js
Comment (h4818) [reply addons/account/static/src/js/section_and_note_fields_backend.js:23]: https://github.com/odoo/odoo/blob/master/addons/web/static/src/views/list/list_renderer.js#L191
No list renderer, no section and note 🤷

File: addons/account/static/src/js/vat_activity.js
Comment (h4818) [reply addons/account/static/src/js/vat_activity.js:12]: That shouldn't be  necessary here, it's not possible to modify `props.value` here.

**Comment by h4818** [comment]
@odoo/rd-security could you please approve moving the file that uses `.html`.

**Review by sdegueldre — APPROVED** [review]
Code quality looks good to me, change in web seem fine, curious about the change in base though.

**Comment by h4818** [comment]
> Code quality looks good to me, change in web seem fine, curious about the change in base though.

Ah it seems this has been resolved and no longer causing the crash. We can probably drop the commit @Levizar

**Comment by smetl** [comment]
@robodoo rebase-ff

File: addons/account/static/src/js/open_move_widget.js
Comment (william-andre) [inline addons/account/static/src/js/open_move_widget.js]: Wouldn't this ideally be a generic field widget 
* on m2o fields
* with a parameter for the function to use on the related model

File: addons/account/static/src/js/vat_activity.js
Comment (william-andre) [inline addons/account/static/src/js/vat_activity.js]: File is named `vat_activity` but the class name is more generic (I think it should be more generic, AFAIK it is also used for other things i.e. for fiduciaries)

File: addons/account/static/src/xml/vat_activity.xml
Comment (william-andre) [inline addons/account/static/src/xml/vat_activity.xml]: *Not related to this, but I noticed it while reviewing this:*

The style on the dashboard seems even more fcked up than in master: all the content of the cards seems pushed to the bottom
image

File: addons/account/static/src/js/account_move_form.js
Comment (william-andre) [inline addons/account/static/src/js/account_move_form.js]: It flckers when you switch tabs.
You need to wait for the save to be done and the changes to be done in the form before switching tabs (was `await this.reload();` in the legacy controller)

File: addons/account/static/src/js/tax_totals.js
Comment (william-andre) [inline addons/account/static/src/js/tax_totals.js:72]: The behavior is weird when you input a non valid float
It will display the old value in the field, but it will tell you that the value is wrong when you try to save the record

**Comment by smetl** [comment]
@robodoo r+

File: addons/account/static/src/js/tours/account.js
Comment (william-andre) [inline addons/account/static/src/js/tours/account.js:62]: All the tours "fix" should go in the commit that broke them

File: addons/account/tests/test_tour.py
Comment (william-andre) [inline addons/account/tests/test_tour.py:27]: > [IMP] account: make tour inherit our test 

Why is this changing?
Have you tried the single l10n build?


<sub>I would be really glad if we could remove this though</sub>

**Comment by william-andre** [comment]
@robodoo r-

**Review by william-andre — COMMENTED** [review]
Nice job :+1: 

The main points are
- [x] the style of the invoice line table is broken
- [x] the drag and drop seems to not work properly anymore (always vendor bills)
- [x] switching tabs on the invoice flickers
- [ ] the fixes of the tests should be in the related commits, not separate
- [ ] the change of the tests, regarding l10n. It deserves a better explanation if we keep it. 

These should be fixed before merging.

There are also other small fixes (minor styling), questions,... that could be done later

File: addons/account/tests/test_tour.py
Comment (smetl) [reply addons/account/tests/test_tour.py:27]: @william-andre This doesn't make sense to have this running for all localizations because some of them needs extra things  by the way.

Comment (william-andre) [reply addons/account/tests/test_tour.py:27]: This is the test for the onboarding tour, that is used in every l10n. Whether you want it or not, it **has to** run for every l10n.

Comment (h4818) [reply addons/account/tests/test_tour.py:27]: This change is not necessary for this branch, we can drop the commit.

Comment (smetl) [reply addons/account/tests/test_tour.py:27]: @william-andre If it's run for the onboarding tour, I agree with you then

File: addons/account/static/src/xml/account_payment.xml
Comment (Levizar) [reply addons/account/static/src/xml/account_payment.xml]: `vertical-align: top;` on the tr element

File: addons/account/static/src/js/bills_upload.js
Comment (Levizar) [reply addons/account/static/src/js/bills_upload.js]: todo: create a new BillListController

File: addons/account/static/src/js/open_move_widget.js
Comment (Levizar) [reply addons/account/static/src/js/open_move_widget.js]: Not to be done in this PR.
to put on backlog

File: addons/account/static/src/js/vat_activity.js
Comment (Levizar) [reply addons/account/static/src/js/vat_activity.js]: rename to JournalDashBoardActivity

File: addons/account/static/src/xml/vat_activity.xml
Comment (Levizar) [reply addons/account/static/src/xml/vat_activity.xml]: Not sure if caused by OWL or by our changes.

File: addons/l10n_ar/static/src/js/tours/account.js
Comment (Levizar) [inline addons/l10n_ar/static/src/js/tours/account.js:13]: Add the FIXME owl comment:
```
        // FIXME WOWL: this selector needs to work in both legacy and non-legacy views
        // because account_invoice_extracts *adds* a js_class on the base view which forces
        // the use of a legacy view in enterprise only
```

File: odoo/addons/test_main_flows/static/tests/tours/main_flow.js
Comment (Levizar) [inline odoo/addons/test_main_flows/static/tests/tours/main_flow.js:993]: add a version for legacy + the FIXME OWL Comment

File: addons/account/static/src/js/tours/account.js
Comment (Levizar) [reply addons/account/static/src/js/tours/account.js:62]: It's fixing the breaks OWL created and we just discover it now because we activate OWL

File: addons/account/static/src/xml/section_and_note_fields_backend.xml
Comment (h4818) [reply addons/account/static/src/xml/section_and_note_fields_backend.xml]: Welcome to Owl

File: addons/account/static/src/js/bills_upload.js
Comment (h4818) [reply addons/account/static/src/js/bills_upload.js]: Try an Image,  everything is fine -> I noticed this when uploading PDF's exported from Odoo

File: addons/account/static/src/js/open_move_widget.js
Comment (h4818) [reply addons/account/static/src/js/open_move_widget.js]: To be done in the future

File: addons/account/static/src/js/bills_upload.js
Comment (Levizar) [reply addons/account/static/src/js/bills_upload.js]: it's a current bug in stable too

File: addons/account/static/src/js/tax_totals.js
Comment (h4818) [reply addons/account/static/src/js/tax_totals.js:72]: bug

File: addons/account/static/src/js/section_and_note_fields_backend.js
Comment (h4818) [reply addons/account/static/src/js/section_and_note_fields_backend.js]: replaced all the `t-call`s

File: addons/account/static/src/js/bills_upload.js
Comment (h4818) [reply addons/account/static/src/js/bills_upload.js]: New Controllers created

File: addons/account/static/src/js/account_move_form.js
Comment (h4818) [reply addons/account/static/src/js/account_move_form.js]: resolved

File: addons/account/static/src/js/tax_totals.js
Comment (h4818) [reply addons/account/static/src/js/tax_totals.js]: fixed

File: addons/account/static/src/xml/account_payment.xml
Comment (h4818) [reply addons/account/static/src/xml/account_payment.xml]: All styles inside the template need to be removed to a scss file...
will be done in another PR

File: addons/account/static/src/xml/account_payment.xml
Comment (h4818) [reply addons/account/static/src/xml/account_payment.xml]: All styles inside the template need to be removed to a scss file...
will be done in another PR

File: odoo/addons/test_main_flows/static/tests/tours/main_flow.js
Comment (h4818) [reply odoo/addons/test_main_flows/static/tests/tours/main_flow.js]: added a comment, and testing legacy and owl

File: addons/l10n_ar/static/src/js/tours/account.js
Comment (h4818) [reply addons/l10n_ar/static/src/js/tours/account.js:14]: added a comment

**Comment by h4818** [comment]
@odoo/rd-security I need another approval here: I need to change the tag name of the generated owl template (generated by `form_compiler`) and just reuse the existing slot contents. There is no way anyone can inject changes here.
@odoo/rd-framework-js is this ok for you?

File: addons/account/static/src/js/section_and_note_fields_backend.js
Comment (Levizar) [reply addons/account/static/src/js/section_and_note_fields_backend.js]: Changing this was probably not worth.
In order to make cleaner extension, we now have a lot more code + a trigger of the security team.
I wonder if it wouldn't be better to just prefix the property with `accountAccountant` to be sure it's never written over.

The mechanics will remains ugly (extending an existing template for everyone) but it avoids a lot weird of code.

File: addons/account/tests/test_tour.py
Comment (Levizar) [reply addons/account/tests/test_tour.py]: commit dropped

File: addons/account/static/src/xml/account_payment.xml
Comment (Levizar) [reply addons/account/static/src/xml/account_payment.xml]: image
It seems to be that way in master too

File: addons/account/static/src/components/account_move_form/account_move_form.js
Comment (h4818) [reply addons/account/static/src/components/account_move_form/account_move_form.js]: I also scroll back to the current position now

File: addons/account/static/src/xml/vat_activity.xml
Comment (h4818) [reply addons/account/static/src/xml/vat_activity.xml]: Fixed

**Comment by william-andre** [comment]
@robodoo r+

**Comment by pedrobaeza** [comment]
Hello, sorry for being late to the party. There's a historical claim that you said to implement that is:

https://github.com/odoo/odoo/pull/63338#issuecomment-821035897

but this never got to be merged. Maybe you can consider this now that on the conversion to OWL?

cc @FlorianGilbert @oco-odoo 

Thanks.

**Pull request closed without merging**
